### ISSUE TITLE
warning: Quasi-quoting might make incremental compilation very inefficient: NtIdent(..)

### ISSUE THREAD

**Issue by SimonSapin** [rust-lang/rust#40946]
I have a macro like this:

```rust
#[macro_export]
macro_rules! define_proc_macros {
    (
        $(
            $( #[$attr:meta] )*
            pub fn $proc_macro_name: ident ($input: ident : &str) -> String
            $body: block
        )+
    ) => {
        $(
            $( #[$attr] )*
            #[proc_macro_derive($proc_macro_name)]
            pub fn $proc_macro_name(derive_input: ::proc_macro::TokenStream)
                                    -> ::proc_macro::TokenStream {
                let $input = derive_input.to_string();
                let $input = $crate::_extract_input(&$input);
                $body.parse().unwrap()
            }
        )+
    }
}

pub fn _extract_input(derive_input: &str) -> &str {
    // …
}
```

When I use it, I get the warning in the title *twice* for every generated functions. When I move things into one crate to avoid "note: this error originates in a macro outside of the current crate", the span of the warning is on `$proc_macro_name` in `#[proc_macro_derive($proc_macro_name)]`.

Steps to reproduce: run`CARGO_INCREMENTAL=1 cargo +nightly build -p cssparser-macros` in https://github.com/servo/rust-cssparser/.

I don’t understand this warning and what I can do about it. What does quasi-quoting mean in this context? How bad is the inefficiency really? Is it possible to fix it while keeping this macro’s flexibility?

CC @michaelwoerister and @nikomatsakis for #37787 that added this warning.

**Comment by michaelwoerister** [comment]
That's strange. @jseyfried, do you see anything in the macro above that in a `Token::Interpolated`?
This is where the warning is generated by the way:
https://github.com/rust-lang/rust/blob/5309a3e31d88def1f3ea966162ed4f81f161d500/src/librustc_incremental/calculate_svh/svh_visitor.rs#L1068

**Comment by michaelwoerister** [comment]
ping @jseyfried: Do you know what could be going on here? Where does the macro definition above introduce a `Token::Interpolated`?

**Comment by jseyfried** [comment]
The `$proc_macro_name` turns into a `Nonterminal::Ident` token. Normally, nonterminals are parsed into AST and don't cause warnings. However, now that #40346 has landed, attributes can have arbitrary tokens and the `Nonterminal::Ident` token ends up in the final AST/HIR.

I believe the solution here is to expand interpolated/nonterminal tokens into their underlying terminal tokens as we lower to HIR.

**Comment by SimonSapin** [comment]
@jseyfried If I understand correctly, something *is* inefficient but what needs to change to fix it is in rustc rather than in my code?

**Comment by michaelwoerister** [comment]
@SimonSapin That is correct. It's an inefficiency in rustc's handling of some parts of the HIR/AST.

**Comment by michaelwoerister** [comment]
> I believe the solution here is to expand interpolated/nonterminal tokens into their underlying terminal tokens as we lower to HIR.

@jseyfried, that sounds like a great solution. Would that be much work?

**Comment by jseyfried** [comment]
@michaelwoerister 
Once #40939 lands it should be simple.

**Comment by rushmorem** [comment]
I'm getting this warning too when using [proc-macro-hack](https://github.com/dtolnay/proc-macro-hack) (cc @dtolnay). Is there any way to suppress it?

```rust
warning: Quasi-quoting might make incremental compilation very inefficient: NtIdent(..)
  --> derive/src/lib.rs:9:1
   |
9  |   proc_macro_expr_impl! {
   |  _^ starting here...
10 | |     pub fn args_impl(input: &str) -> String {
11 | |         args::Args::new(input)
12 | |             .process().tokens
13 | |             .to_string()
14 | |     }
15 | | }
   | |_^ ...ending here
   |
   = note: this error originates in a macro outside of the current crate
```

**Comment by seanmonstar** [comment]
If it will take a while to fix this, it would be nice to at least be able to suppress the warning.

**Comment by michaelwoerister** [comment]
@jseyfried Is there an ETA for #40939?

**Comment by michaelwoerister** [comment]
Now that https://github.com/rust-lang/rust/pull/40939 has landed, will you look into a fix, @jseyfried?

**Comment by jseyfried** [comment]
@michaelwoerister Sure, I still think https://github.com/rust-lang/rust/issues/40946#issuecomment-296570557 is the best fix; I'll implement this week.

**Comment by michaelwoerister** [comment]
Thanks a lot, @jseyfried ❤️

**Comment by nikomatsakis** [comment]
@jseyfried any updates on status of this?

**Comment by alexcrichton** [comment]
I've made a PR for this at https://github.com/rust-lang/rust/pull/44601, although @jseyfried if that's not at all what you were thinking, just let me know!
